File: csharp/parsing-log-files/ParsingLogFiles.cs
using System.Text.RegularExpressions;

public class LogParser
{
    private string isLogPattern = @"^\[(TRC|DBG|INF|WRN|ERR|FTL)\].+$";
    private string splitPattern = @"\<[\^\*=-]*\>";
    private string pwdPattern = @"(?i)\""password\""";
    private string eofPattern = @"end-of-line\d+";
    private string weakPattern = @"(?i)(\bpassword\w+\b?)";

    public bool IsValidLine(string text) => new Regex(isLogPattern).IsMatch(text);

    public string[] SplitLogLine(string text) => Regex.Split(text, splitPattern);

    public int CountQuotedPasswords(string lines) => Regex.Split(lines, pwdPattern).Count();

    public string RemoveEndOfLineText(string line) => Regex.Replace(line, eofPattern, string.Empty);

    public string[] ListLinesWithPasswords(string[] lines) =>
        lines.Aggregate([], (string[] acc, string line) => [.. acc, Regex.Matches(line, weakPattern).Aggregate($"--------: {line}", (string acc, Match m) => $"{m.Value}: {line}")]);
}

File: go/parsing-log-files/parsing_log_files.go
package parsinglogfiles

import "regexp"

func IsValidLine(text string) bool {
	re := regexp.MustCompile(`^\[(TRC|DBG|INF|WRN|ERR|FTL)\]`)
	return re.MatchString(text)
}

func SplitLogLine(text string) []string {
	re := regexp.MustCompile(`<[~*=\-]*>`)
	return re.Split(text, -1)
}

func CountQuotedPasswords(lines []string) int {
	re := regexp.MustCompile(`([^"]*password[^"]*)`)
	count := 0
	for _, line := range lines {
		count += len(re.FindStringSubmatch(line))
	}
	return count
}

func RemoveEndOfLineText(text string) string {
	re := regexp.MustCompile(`end-of-line\d+`)
	return re.ReplaceAllLiteralString(text, "")
}

func TagWithUserName(lines []string) []string {
	re := regexp.MustCompile(`User\s+(\S+)`)
	result := make([]string, len(lines))

	for i, line := range lines {
		matches := re.FindStringSubmatch(line)
		if len(matches) > 1 {
			username := matches[1]
			result[i] = "[USR] " + username + " " + line
		} else {
			result[i] = line
		}
	}

	return result
}
